### PR TITLE
Specialized copy! for sparse matrices and vectors

### PR DESCRIPTION
**Pull request by martinholters**
This implements specialized `copy!` for sparse matrices of different sizes and sparse vectors, significantly improving performance.

### CODE REVIEW

**Comment by tkelman** [comment]
Very nice! For some reason Travis didn't run here. Once AppVeyor finishes, closing then reopening the PR should trigger Travis.

**Comment by martinholters** [comment]
Closing...

**Comment by martinholters** [comment]
...and reopening as suggested to trigger Travis.

**Comment by KristofferC** [comment]
Nice that this got properly done.

File: base/sparse/sparsematrix.jl [hunk 1]
```diff
@@ -193,20 +193,51 @@ end
...
+        end
     else
-        invoke(Base.copy!, Tuple{AbstractMatrix{TvA}, AbstractMatrix{TvB}}, A, B)
+        length(A) >= length(B) || throw(BoundsError())
+        lB = length(B)
+        nnzA = nnz(A)
+        nnzB = nnz(B)
+        # Up to which col, row, and ptr in rowval/nzval will A be overwritten?
+        lastmodcolA = div(lB - 1, A.m) + 1
+        lastmodrowA = mod(lB - 1, A.m) + 1
```
Comment (timholy) [inline base/sparse/sparsematrix.jl]: There's also `divrem`, if it makes sense. Or just work with `lB - lastmodcolA*A.m`.

**Comment by timholy** [comment]
Just skimmed it briefly, but it seems like good work with extensive tests. Nice. How big is the speedup?

**Comment by timholy** [comment]
That change I suggested isn't important enough to have to go through a whole new round of CI, unless you have other changes pending.

**Comment by martinholters** [comment]
The "resize but keep elements at end instead of beginning" part could be DRYed but I don't think it's worth it (and I have no good idea how to name that function). So I have no changes planned.

Speedup of course depends on a lot of things, so I didn't do any real benchmarking. In the ideal case - no resizing of `nzval` necessary - I think I saw a factor of 1000 for a 20x50 matrix with 20% non-zero entries.

**Comment by timholy** [comment]
Thanks for an excellent contribution!

File: base/sparse/sparsematrix.jl [hunk 2]
```diff
@@ -193,20 +193,51 @@ end
...
         resize!(A.rowval, length(B.rowval))
-        copy!(A.rowval, B.rowval)
-        copy!(A.nzval, B.nzval)
+        if size(A) == size(B)
+            # Simple case: we can simply copy the internal fields of B to A.
+            copy!(A.colptr, B.colptr)
+            copy!(A.rowval, B.rowval)
+        else
+            # This is like a "reshape B into A".
+            sparse_compute_reshaped_colptr_and_rowval(A.colptr, A.rowval, A.m, A.n, B.colptr, B.rowval, B.m, B.n)
```
Comment (tkelman) [inline base/sparse/sparsematrix.jl]: this function should probably have an `!` on it

**Pull request merged**